### ISSUE TITLE
[css-cascade] [css-nesting] Figure out whether we're fine with "shifting up" bare declarations after rules

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#8738]
If you do:

```
div {
  color: green;
  @media (width > 0) {
    color: red;
    background: red;
  }
  background: green;
}
```

My understanding is that per spec the div color and background would be red.

That seems rather confusing. There are various alternatives here:

 * We're ok with this.
 * We forbid declarations after nested rules.
 * We deal with bare declarations by somehow sorting them together, so that would be effectively something like:

```
div {
  color: green;
  @media (width > 0) {
    color: red;
    background: red;
  }
  & {
    background: green
  }
}
```

Maybe something else?

cc @fantasai

**Comment by Loirooriol** [comment]
> We forbid declarations after nested rules

Seems potentially problematic that some people may have some garbage followed by lots of declarations, then at some point we add some feature that parses the garbage as a nested rule, and then all the declarations stop applying. Spooky action at distance.

One of the main characteristics of option 3 (later modified with lookahead) is that it allows freely mixing declarations and nested rules. So at this point I don't think it makes sense to restrict that. The proper way to have such restriction would have been choosing option 4 or similar (which BTW seemed better to me).

**Comment by tabatkins** [comment]
Right, I argued in previous discussions that forbidding decls after rules is fundamentally problematic. The basic question is "when has a rule occurred?", and this needs to *not* be dependent on whether a rule is valid or not, as that would mean different browser levels would interpret following properties differently.

So you need to define some notion of when we've found a rule, that allows for invalid rules, and do so in a way that doesn't unduly restrict our future syntax options.

For example, if we ever allow {} in a property, and you write it invalidly so it triggers rule parsing, would that kick the "now there's a rule" switch? Does that mean we actually have to *forbid* ever using `{}` in properties?

------

On the other hand, sticking with the current design that just allows them, and relying on people to not write unreadable stylesheets, suffers from none of these issues.

**Comment by emilio** [comment]
Not saying it's necessarily a good idea, but playing devil's advocate a bit, we already have that kind of concept for e.g. `@import` / `@namespace` / etc, don't we?

Imagine someone wrote this ten years ago:

```css
:has(body) {}

@import "something.css";
```

(Or something along those lines)

Their `@import` would stop working on browsers that support `:has()`, yet that hasn't been ever a concern (and I don't think it should be).

The switch for that is "A valid rule has been parsed", and I think we could do the same, so that random garbage doesn't cause that action at a distance.

**Comment by tabatkins** [comment]
I think these are pretty different situations in practice, tho. We add to the set of "things that can go before `@import`" super rarely -- in fact, we've only done it once, with `@layer`.

On the other hand, we add new things that would qualify for nesting fairly regularly - we recently added `@scope`, and just resolved to add `@initial-name-tbd`. So the set of things that may or may not trigger "no more properties from here on" is regularly changed, presenting new hazards.

Ultimately, the question is still - what are we trying to protect authors from? This sort of interleaving has been allowed in Sass and related tools for many years, with the exact same "pretend all the properties are together at the top" behavior, and never caused notable issues.

**Comment by SebastianZ** [comment]
> For example, if we ever allow {} in a property, and you write it invalidly so it triggers rule parsing, would that kick the "now there's a rule" switch? Does that mean we actually have to _forbid_ ever using `{}` in properties?

If you mean the _property value_ and I interpret the algorithm correctly, then `{}` is already allowed by consuming a [component value](https://drafts.csswg.org/css-syntax-3/#consume-a-component-value) in step 5 of [consuming a declaration](https://drafts.csswg.org/css-syntax-3/#consume-a-declaration). And it can't trigger rule parsing because the consumed component value is appended to the declaration's value.

Sebastian

**Comment by tabatkins** [comment]
This is in reference to the resolved-on new parsing algo that tries to parse as a decl, then falls back to parsing as a rule if the result wasn't valid. That algo isn't in the Syntax spec quite yet.

**Comment by cdoublev** [comment]
> That algo isn't in the Syntax spec quite yet.

Should we expect more changes to come (in CSS Syntax 3, at least) after this [commit](https://github.com/w3c/csswg-drafts/commit/6ab5888d05ef3f445e26a1ca81081a7b912016ed)?

`color: red; @media { foo: bar } color: green` does not seem to parse as I would expect against `<declaration-list>`, probably because there is no `;` after `foo: bar`. I think `}` should be provided to [consume a declaration](https://drafts.csswg.org/css-syntax-3/#consume-declaration).

**Comment by tabatkins** [comment]
Thanks for spotting that! I hadn't gotten around to updating my own parser to the new text to test it, so I missed that I was mishandling nested constructs.

I believe it's all good now; the three construct consumption algorithms (at-rule, qualified rule, declaration) now all take a "nested" bool, which triggers them to stop early when encountering a top-level unmatched `}`. This is passed by "parse a block's contents" (which is renamed from "parse a declaration list" since it definitely returns more than just declarations).

**Comment by cdoublev** [comment]
I do not mind waiting for an update to your parser to validate your updates. This would prevent me from polluting this issue with implementation details.

That said, `<declaration-list>` could be renamed to `<statement-list>` and could be parsed with *consume a list of statements*.

This would allow using `<declaration-list>` for when a list of declarations (strict) is required (eg. keyframe rule, `@font`, etc). But I do not know if it would be backward compatible to apply it to existing rules, and your intent may even be to preserve this flexibility to accept declarations/rules within any rule, for future extensibility.

**Comment by tabatkins** [comment]
I've done the parser update, but had to stop at end of day before I could get my testing framework updated to the new data structures. Later today I'll have it working. ^_^

> But I do not know if it would be backward compatible to apply it to existing rules, and your intent may even be to preserve this flexibility to accept declarations/rules within any rule, for future extensibility.

Yes, *all* rules already handled at-rules in their blocks even if they only accepted declarations validly, so that's staying, and at that point there's not really any reason to continue having a parsing split. All blocks are parsed the exact same way now in Syntax, accepting all three kinds of constructs (at rules, qualified rules, declarations).

My plan on reshuffling the productions is just to define one generic production, and then *probably* some sub-productions that automatically imply certain restrictions on what's valid so you don't have to say it in prose. But they won't change the parsing behavior.

**Comment by LeaVerou** [comment]
Unfortunately, it appears that Chrome shipped before we could decide on this, and now authors are already [blogging about the "gotcha"](https://kilianvalkhof.com/2023/css-html/the-gotchas-of-css-nesting/).

Hopefully it's not too late to change this. 
I think we should *really* try to avoid any rewriting that changes order of declarations.

Since we’re already [resolved that nested MQs will wrap their contents with an implicit `& {}`](https://github.com/w3c/csswg-drafts/issues/7850#issuecomment-1497794190), perhaps we can do the same thing with any declarations after the first nested rule, as @emilio proposes in the OP?

**Comment by tabatkins** [comment]
I still feel very strongly that we should not change this, and the current behavior is the best. Again, this is the *exact* behavior that Sass (and I suspect other preprocessors) have had for a decade+ already, and it has not been a problem there. (Largely because people *just don't write that code* - they put their declarations first, *then* their nested rules.) I don't think we should try to add more behavior for something that has proven itself to not be a problem in practice.

MQs wrap *all* naked properties in an `& {}`; style rules obviously *cannot* do this, so the behavior would be inconsistent either way. As the spec is currently written, style rules and MQs are each internally consistent with a single behavior for all naked properties. I think it would be a (probably minor) bad thing for style rules to have two behaviors, depending on relative ordering of rules and declarations.

**Comment by kizu** [comment]
While I would like to see a world where the order would stay as written, I think I agree with Tab on that this is the behavior that was implemented in all preprocessors: [less](https://lesscss.org/less-preview/#eyJjb2RlIjoiI2xpYigpIHtcbiAgICAuY29sb3JzKCkge1xuICAgICAgQHByaW1hcnk6IGJsdWU7XG4gICAgICBAc2Vjb25kYXJ5OiBncmVlbjtcbiAgICB9XG4gICAgLnJ1bGVzKEBzaXplKSB7XG4gICAgICBib3JkZXI6IEBzaXplIHNvbGlkIHdoaXRlO1xuICAgIH1cbiAgfVxuICBcbiAgLmJveCB3aGVuICgjbGliLmNvbG9yc1tAcHJpbWFyeV0gPSBibHVlKSB7XG4gICAgd2lkdGg6IDEwMHB4O1xuICAgIGhlaWdodDogKCR3aWR0aCAvIDIpO1xuICB9XG4gIFxuICAuYmFyOmV4dGVuZCguYm94KSB7XG4gICAgQG1lZGlhIChtaW4td2lkdGg6IDYwMHB4KSB7XG4gICAgICB3aWR0aDogMjAwcHg7XG4gICAgICAjbGliLnJ1bGVzKDFweCk7XG4gICAgfVxuICB9IiwiYWN0aXZlVmVyc2lvbiI6IjQueCJ9), [sass](https://www.sassmeister.com/) and [stylus](https://runkit.com/kizu/648cbbea05e821000873adde), at least some CSS-in-JS solutions (tested in styled-components, couldn't find a good playground to share), and `nested` PostCSS plugin (can be tested [here](https://sneakertack.github.io/postcss-playground/)), [lightningCSS](https://lightningcss.dev/playground/index.html#%7B%22minify%22%3Atrue%2C%22nesting%22%3Atrue%2C%22customMedia%22%3Atrue%2C%22cssModules%22%3Afalse%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22chrome%22%3A6225920%7D%2C%22include%22%3A0%2C%22exclude%22%3A0%2C%22source%22%3A%22div%20%7B%5Cn%20%20color%3A%20green%3B%5Cn%5Cn%20%20%40media%20(min-width%3A%200)%20%7B%5Cn%20%20%20%20color%3A%20red%3B%5Cn%20%20%20%20background%3A%20red%3B%5Cn%20%20%7D%5Cn%20%5Cn%20%20background%3A%20green%3B%5Cn%7D%22%2C%22visitorEnabled%22%3Afalse%2C%22visitor%22%3A%22%7B%5Cn%20%20Color(color)%20%7B%5Cn%20%20%20%20if%20(color.type%20%3D%3D%3D%20'rgb')%20%7B%5Cn%20%20%20%20%20%20color.g%20%3D%200%3B%5Cn%20%20%20%20%20%20return%20color%3B%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22unusedSymbols%22%3A%5B%5D%2C%22version%22%3A%22local%22%7D) (but it polyfills the current spec).

Given this was done literally _everywhere_, I'd say it is not worth it to change this.

**Comment by LeaVerou** [comment]
I thought it was very odd that the [WebKit poll](https://webkit.org/blog/14571/css-nesting-and-the-cascade/) seems to be favoring Option 1, so I posted a couple more polls:

- Twitter/X: https://twitter.com/LeaVerou/status/1709573062411088204
- Mastodon: https://front-end.social/@leaverou/111177156184963593

which so far seem to be showing a _very_ different picture with more than 3 out of 4 expecting no shifting. There are even [people incorrectly explaining](https://front-end.social/@steveworkman@webperf.social) what happens:

> supports doesn't change specificity, so it'd be the last one in the ruleset

I think the current behavior is **extremely unintuitive**, and 10 years down the line we will regret opting for consistency with today’s preprocessors over predictable behavior. In fact, we even have a TAG principle advising against this exact thing: [Prioritize usability over compatibility with third party tools](https://www.w3.org/TR/design-principles/#third-party-tools).
I'd even vote for entirely disallowing declarations after rules over the current behavior, as it would give us more time to figure this out.

Perhaps we need more data here. Maybe an MDN short survey?

**Comment by LeaVerou** [comment]
(fat fingered the close button)

Agenda+ to resolve to temporarily disallow declarations after rules while we try to get better data here, so that we don't get stuck in a situation where we can't change the behavior due to web compat.

**Comment by mdubet** [comment]
FWIW we (WebKit) were also very surprised by our poll results (it actually changed over the weekend, the first days were Option 2 winning 70/30).

Disallow declarations after rules seems the "worst" solution because it means we have to determine the "a rule have been parsed" trigger which could cause some compatibility issue if we extend what correctly parse as a rule. However, I'm not sure what would be the risk if we allow declarations after rules like now and follow the cascade - so last one wins?

**Comment by tabatkins** [comment]
> Could you please elaborate on that? Not sure I follow what you mean by "a rule has been parsed trigger".

Yeah, I've elaborated on this in the past when Emilio suggested disallowing properties after rules.

So, this is a parsing switch. In theory parsing switches are doable; we explored a few of these earlier when working thru Nesting options. But the switch needs to be reliable - the earlier discussion about `@nest;` being the switch worked, because we know exactly what to look for and don't expect that to change in the future.

But if the parsing switch is "a rule of some kind is seen", that's hard. The most obvious interpretation of that is "a *valid* rule of some kind is seen" - that's well-defined, but it means that authors can see unexpected differences in parsing behavior if the rule in question is supported in some browsers but not others, as older browsers will throw out the rule and continue allowing declarations, while newer browsers will see it and disallow declarations. (And consider: an invalid *selector* makes the style rule invalid, and we do new selectors all the time.)

So ideally we define invalid rules as also triggering this. But then we open up the full syntax space of what an "invalid rule" actually *is*. Is `foo bar:baz {...};` an invalid rule? Or is it a new property syntax? We can define something for this, but it means restricting our future evolution capabilities, to a larger extent than what we've already done for Nesting.

**Comment by b-strauss** [comment]
> FWIW we (WebKit) were also very surprised by our poll results (it actually changed over the weekend, the first days were Option 2 winning 70/30).

Have the results been restricted? I remember seing opposite numbers yesterday.

**Comment by jensimmons** [comment]
I watched the results of the poll very carefully in the first couple days. It was consistently 37-39% for Option 1, and 61-63% for Option 2. 

It's very telling when you hit enough votes (like 100) and the results stabilize. As more and more votes came in, the results did not change. 

Sept 28 at 2:50pm ET.
<img width="294" alt="Screenshot 2023-09-28 at 2 50 18 PM" src="https://github.com/w3c/csswg-drafts/assets/108474/f4bd7565-b89e-41e2-89af-fb50501c1e5a">

Sept 29 at 12:50pm ET.
<img width="306" alt="Screenshot 2023-09-29 at 12 50 02 PM" src="https://github.com/w3c/csswg-drafts/assets/108474/2748ddbb-3590-416e-b240-f7aaec14b2c5">

Then over the weekend, another 1800 votes came in, with an overwhelming preference for Option 1. Almost three times as many votes came in over the weekend? Long after we stopped promoting the article on social media? With a radically different result? 

That's a bot. 

So we decided to close the survey and post the last known results from before the traffic pattern became highly suspect. 

You can see a similar preference for Option 2 in Lea's survey: https://mastodon.social/@leaverou@front-end.social/111177156433448874

**Comment by jensimmons** [comment]
I agree with Lea:
> I think the current behavior is extremely unintuitive, and 10 years down the line we will regret opting for consistency with today’s preprocessors over predictable, natural behavior.

A desire to match Sass is a terrible reason. (Especially if the only reason Sass & other tools made their choice is that they could not implement the more intuitive behavior.)

We should be designing the language for the future — for 20+ years from now, when the majority of developers have never used Sass, and those that did don't remember how it worked. 

I've not heard any other reason put forth besides: this is how all the tooling does it, and we don't think it's a big deal. I don't hear anyone arguing that this design is good for the cascade, or makes sense, or is something future developers will easily learn.

I believe this will be a very confusing problem for developers trying to debug their code if we don't fix it.

**Comment by kizu** [comment]
I think I was initially for option 1, but gradually moved towards option 2.

Mainly, deciding for me are the “designing the language for the future” combined with the case being very rare and usually considered a bad practice by itself. If it was something that is present in all current preprocessors, but very commonly used it would be a different story, but this is a case of an edge-case, where preprocessors decided to interop for some reason.

I think it should be ok to handle this edge-case differently, but properly.

**Comment by LeaVerou** [comment]
Well said @jensimmons. I would urge anyone who wants to weigh in on this to also read the responses to both of the polls I posted. They are even more illuminating than the (sweeping) quantitative data.

**Comment by jensimmons** [comment]
I do not believe it's correct for these two blocks to end up with different results:
```
h1 {
  color: yellow;
  @media (width > 0) {
    color: red;
  }
  color: green;
}
```
```
h2 {
  color: yellow;
  @media (width > 0) {
    color: red;
  }
  & {
    color: green;
  }
}
```
(Try it in a browser with support for CSS Nesting: https://codepen.io/jensimmons/pen/KKbrJBp/5861d875920bb25695c12975bf627b75?editors=1100 )

The ampersand should be a clean substitute for the unnested selector, not something that changes the result.

**Comment by ydaniv** [comment]
I've used Less a lot, and have been using Sass for over 6 years now on a huge strictly Sass code base, and I've never stumbled on this behavior, I'm actually surprised by it. I guess *everyone* have been writing declarations before rules everywhere.

So IMHO this is also counter intuitive for most devs whom are using preprocessors, that expect the expected CSS behavior of "latter wins".

**Comment by Loirooriol** [comment]
I think it's clear that the current behavior is quite bad because it breaks the principle of "last declaration (with same specificity) wins", I'd have objected against option 3 if I had realized this at the beginning. But at this point we are stuck with that syntax, and trying to address this problem within option 3 seems to cause even worse outcomes, so probably we will have to live with it. Could be added to the list of CSS mistakes.

**Comment by tabatkins** [comment]
> (Especially if the only reason Sass & other tools made their choice is that they could not implement the more intuitive behavior.)

They absolutely *could* have implemented either behavior; it's just outputting a separate rule rather than combining into one rule. They do exactly that if you *do* wrap the latter declaration in a `& {...}` rule, so both behaviors are clearly possible.

> I do not believe it's correct for these two blocks to end up with different results:

I think the consistency argument is reasonable in either direction. Before nesting, if you wrote:

```css
h1 {
  color: yellow;
  color: green;
}
```

you'd get a single 'color' declaration with the value green. One can reasonably argue, I think, that it's also consistent that adding an unrelated rule (the `@media` in your example) shouldn't change the behavior of these declarations. I suspect that might be why the preprocessors originally chose the behavior that they did.  Adding an explicit `& {...}` wrapper around the latter declaration is a much stronger declaration of intent than just inserting an unrelated rule before it.

------

I have no strong opinion on which way we go for this. But the fact that the current spec *is* the behavior of essentially every preprocessor, and afaict there have been approximately zero complaints about it for over a decade of use (because, again afaict, nobody actually writes code like that in the first place), means that there's very little reason for *us* to care about what the behavior is either. As such I'd prefer no change, as compatibility with the wider ecosystem is a (minor) benefit, but I won't object over the rest of the WG if the decision goes the other way.

**Comment by LeaVerou** [comment]
@tabatkins 

If preprocessor users haven’t really stumbled on this in the first place, compatibility with preprocessors is not a benefit, minor or otherwise. It's only a benefit when it's compatible with behaviors they have, actually, experienced.
Adding something to CSS is a much wider deployment than adding it to a preprocessor, so “people haven't hit this problem before” should not be an excuse for weird behavior.

If you read through the thread and the various polls, there is a very strong signal from developers that the current behavior is confusing. Even worse, for the few that don't find it confusing, it’s due to a broken mental model about the cascade: they thought that `@media` *adds* specificity. So I’m quite worried not just about the ergonomics of this, but also what it teaches authors about the rest of CSS.

And it's not like there's an actual implementation reason for the confusing behavior, right? It seems we all agree (?) that changing it produces better ergonomics. So what's the argument for keeping it? Compatibility with Sass and co? We literally have a TAG principle about this exact thing: [2.12 Prioritize usability over compatibility with third-party tools](https://w3ctag.github.io/design-principles/#third-party-tools).

**Comment by emilio** [comment]
The argument for keeping it could be performance, mostly. E.g., if you do something like:

```css
.foo {
  --bar: baz;
  @media (a) {
    --bar: something-else;
  }

  --baz: ...;
  @media (b) {
    --baz: something-else;
  }

  // Repeat x100 etc
}
```

If we generate a bunch of split rules for anything after an `@media` rule, that can cause useless overhead, which is also surprising.

**Comment by mdubet** [comment]
Implementation wise, I think it's a bit more complicated to implement the "last declaration (after rules) wins" if we assume the most obvious way : wrap following declarations in `& { }`

- the serialisation will generally be more verbose
- it means in a future browser some declarations might be wrap in a `& { }`  while they are not currently (because some garbage becomes a valid rule) (I doubt it's an actual issue though?)
- the pseudo-element issue cf snippet will need some additional treatment (until we actually fix the underlying issue and just allow pseudo-element matching for `:is()` and `&` (and it's already a weirdness so maybe we don't care neither).

```html
<style>
.foo::before {
  background-color: blue;
  content: 'nonest';
}
.foo::before {
 & {
  background-color: green;
  content: 'nest';
  }
}
</style>

<div class=foo>Foo</div>
```

There might be a better way to implement all this though.

**Comment by LeaVerou** [comment]
@emilio 
> An argument for keeping it could be performance, mostly. E.g., if you do something like:
> [snip]
> If we generate a bunch of split rules for anything after an `@media` rule, that can cause useless overhead, which is also surprising.

Given that this doesn’t happen much, I don’t think that would be significant in practice. Also, I suspect when people write code that way (with a declaration, and then a MQ after it to set just that property), the MQs are not that varied, so a low-hanging optimization would be to merge them together when it doesn’t change the outcome. Meaning:

```css
.foo {
	grid-template-columns: auto 1fr auto;
	@media (width < 500px) {
		grid-template-columns: auto 1fr;
	}
	
	gap: .5em;
	@media (width < 500px) {
		gap: .3em;
	}
}
```

becomes:

```css
.foo {
	grid-template-columns: auto 1fr auto;
	gap: .5em;

	@media (width < 500px) {
		grid-template-columns: auto 1fr;
		gap: .3em;
	}
}
```

This is also consistent with our rule to [serialize to the shortest equivalent syntax](https://github.com/w3ctag/design-principles/issues/284#issuecomment-773611014). 

@mdubet 
> the serialisation will generally be more verbose

Not if we go the other way and decide to serialize *all* `& {}`s after rules as bare declarations. Which would also be consistent with our rule to favor the shortest equivalent syntax on serialization. This would also be more likely to preserve author intent, since I bet most `& {}` rules would have come from wrapping bare declarations, rather than specified explicitly by the author.

**Comment by emilio** [comment]
I don't think we have a precedent for rewriting / merging rules, and I don't think we'd want to add such thing.

**Comment by LeaVerou** [comment]
> I don't think we have a precedent for rewriting / merging rules, and I don't think we'd want to add such thing.

Wrapping declarations in `& {}` is also technically rewriting, so I think that ship may have sailed?

**Comment by mirisuzanne** [comment]
I wonder if we should consider how this behavior relates to similar future features, e.g. [mixins](https://github.com/w3c/csswg-drafts/issues/9350), if we eventually go down that path. For example, an `@apply --x-mixin;` (or similar) rule could be treated as an extension of nesting – a placeholder for `& { <output of --x-mixin> }`. With mixins it is more clearly important that authors have an easy way to override the output by providing additional declarations after the mixin. In fact, most Sass 'best practices' have encouraged putting mixins _before_ declarations rather than after.

Clearly, nesting and mixins wouldn't _need_ to behave the same (they don't in Sass), but it seems like a useful comparison to consider?

**Comment by tabatkins** [comment]
@leaverou

> If preprocessor users haven’t really stumbled on this in the first place, compatibility with preprocessors is not a benefit, minor or otherwise.

Right, but my point is that *neither behavior is a benefit*, according to over a decade of experience. This simply *does not matter to authors*, as far as we can tell. So the "benefit to authors" part of the PoC is approximately 0; at best it's a learnability benefit, but better learnability for a case that doesn't appear to ever happen in practice is worth extremely little. So impl benefits, even minor ones, can weigh sufficiently here to sway the outcome.

> a low-hanging optimization would be to merge them together when it doesn’t change the outcome. 

We cannot do this. What things are raw decls and what are rules is observable in the OM. I absolutely do not want the OM to be dependent on "when it doesn't change the outcome".

> Wrapping declarations in & {} is also technically rewriting, so I think that ship may have sailed?

No, it's not at all rewriting, in the sense that you're suggesting. How we interpret things on the first pass is up to us; you're talking about a *completely different thing* where something would be wrapped in a rule or not depending on the exact property names used across/between *other* rules. (Whether or not the first `@media` in your example used 'gap' would affect whether it "changes the outcome"!) That's an extremely non-local and non-obvious effect, and I'd object strongly to it.

> Not if we go the other way and decide to serialize all & {}s after rules as bare declarations. Which would also be consistent with our rule to favor the shortest equivalent syntax on serialization.

We can't do this either, because it would mean that two consecutive `&{...}` blocks would roundtrip into a single one. And I don't think we want to specify that the serialization differs based on whether there's a single or multiple blocks with a particular selector in a row.


@miriamsuzanne

> With mixins it is more clearly important that authors have an easy way to override the output by providing additional declarations after the mixin. In fact, most Sass 'best practices' have encouraged putting mixins before declarations rather than after.

This is a fair argument, and does argue for a real author benefit in doing the rule-wrapping. Sass gets away with it because it's just doing direct substitution of the `@include` with its contents.

Note that Sass's behavior is inconsistent in this regard; for example:

```css
@mixin reset-list {
    & {
        margin: 0;
        padding: 0;
        list-style: none;
    }
}

ul {
    @include reset-list;
    margin: 1em;
}

/* compiles to */

ul {
  margin: 1em;
}
ul {
  margin: 0;
  padding: 0;
  list-style: none;
}
```

So the "trailing" `margin: 1em` gets put before the mixin in the cascade and will be overridden! (Defining the mixin with naked properties rather than a nested rule does do what you suggested, tho.)

**Comment by tobymackenzie** [comment]
I was able to find [a Sass issue desiring the no "shifting up" behavior](https://github.com/sass/sass/issues/1243) from 2014.  It is still open and tagged as "planned", though seemingly no work has been done on it since.

**Comment by vrubleg** [comment]
The longer this ticket stays forgotten, the harder it will be to change in the future.

**Comment by andruud** [comment]
@vrubleg Not forgotten, I'm investigating the feasibility of this. I had high hopes that we could make this change, but the [use-counter](https://chromestatus.com/metrics/feature/timeline/popularity/4709) I've added so far shows that ~0.16% of pageloads already place bare declarations after nested rules. This is too much to make a potentially breaking change.

I'm now working on making a tighter use-counter, as spot checks of actual sites make me believe there often isn't an actual contest between the bare declarations and the nested rules (e.g. they use non-overlapping sets of properties).

**Comment by SelenIT** [comment]
Since the existing behavior follows the pre-existed preprocessors logic and authors start to get used to it, wouldn't it be the easiest solution just to add "nesting depth" as an explicit cascading order criterion between Specificity and Order of Appearance?

**Comment by LeaVerou** [comment]
> Since the existing behavior follows the pre-existed preprocessors logic and authors start to get used to it, wouldn't it be the easiest solution just to add "nesting depth" as an explicit cascading order criterion between Specificity and Order of Appearance?

The Cascade is already insanely complicated, and we've reached the point that most authors don't grok it fully. We should not be increasing its complexity further!

**Comment by SelenIT** [comment]
It seems to me that adding one more order criterion to 6-7 existing adds about the same cognitive load as complicating one of the existing criteria with de facto a new implicit reordering rule, but an explicit step in the algorithm would be probably a bit clearer and easier to grok for many people 😳

**Comment by LeaVerou** [comment]
Possibly true, but there's also the option of *not* shifting!

**Comment by Loirooriol** [comment]
Just pointing out that pseudo-elements aren't the only problem, there is also #9806

```html
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<style>
/*<![CDATA[*/
@namespace url("http://example.com/foo");
@namespace svg url("http://www.w3.org/2000/svg");
svg|svg {
  background: red;
  & * { --foo: bar; }
  background: green;
}
/*]]>*/
</style> 
<svg id="svg" width="100" viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg"></svg>
</html>
```

This is currently green in browsers. But if we convert it into

```css
svg|svg {
  background: red;
  & * { --foo: bar; }
  & { background: green; }
}
```

then it's red, because `&` is not a type selector so it can only match in the default namespace, and our element is in the svg namespace.

**Comment by andruud** [comment]
Not sure why this is tagged Agenda+, so this may or may not be relevant, but to follow up from January: improved use-counters are implemented in Chrome 123. Results are not in yet.

**Comment by mdubet** [comment]
It seems that all the proposals try to work around the fact that we can’t mix declarations and rules because the OM CSSStyleRule/CSSGroupingRule doesn’t allow this : so we either shift the bare declarations up to group them (current behaviour) or rewrite the following bare declarations to look like rules (and because we can’t do that for the first group of declarations inside the rule, we try to find a switch point like first-valid-rule or @nest).

What about introducing a new CSSBlockContent (with contains a list of declarations and rules intermixed) ? It would be the content of style rule, @media rule, @container rule, @scope rule.. - all nested group rules https://drafts.csswg.org/css-nesting/#nested-group-rules

For backward compatibility, CSSStyleRule would derive from it but would also maintain the current CSSStyleDeclaration API behaviour - overwriting similar declarations and ignoring interleaved rules ; but with the new behaviour visible through the CSSBlockContent API to get the list of rules/declarations in the actual cascade respected order.

It introduces some complexity for implementation, but seems doable ?

**Comment by LeaVerou** [comment]
I think something like `@nest` could work provided that:
- It gets generated in the background when authors use declarations after rules
- When serializing an `@nest` rule, it gets serialized as bare declarations.

Then essentially it becomes an esoteric implementation detail, that only those doing specialized CSSOM handling need to learn.

**Comment by emilio** [comment]
@mdubet can you elaborate on how your proposal would look like? In general mixing rules and declarations seems probably-fine but, at least implementation wise, it's a bit more complicated because:

 * Some things still need just declarations at the end of the day (like the style attribute, SMIL, mapped attributes).
 * All the relevant code is super-hot, so I'm a bit wary of the performance impact of introducing a lot of branches and complexity there.

Another issue is that a lot of the algorithms in the CSSOM don't work if there are mixed rules. For example, consider:

```css
foo {
  padding-top: 0;
  padding-left: 0;
  & { padding-top: 10px; }
  padding-bottom: 0;
  padding-right: 0;
}
```

If there wasn't a nested rule in the middle, the OM would serialize `padding: 0px;`. (To be clear, it doesn't seem hard to fix, just another thing to think about).

In any case I'm curious about how you're envisioning it, because that doesn't seem super clear. I assume you want something like:

```webidl
interface CSSBlockContent : CSSStyleDeclaration {}
```

Or such, where `CSSStyleRule.prototype.style` would start returning a `CSSBlockContent`? One quirk of such interface is that `CSSStyleDeclaration` is "map-like" right now, and iterable (it yields the longhands that are present), and it's not totally clear to me how you'd expose the mixed rules somewhat sanely...

**Comment by mdubet** [comment]
For implementation, both because as you said some things still only need declarations and because the whole codebase is kinda built/optimized around the concept of "1 rule = 1 selector + 1 array of declaration", we would need - in WebKit at least - to separate the rule into pieces at parse time (each block of declarations and each rule being a piece) and keep the original order of appearance in memory (because it's not exactly the implicit order of rules anymore). 

It should just have a performance impact at parsing time to split the rule, during style resolving for order of appearance (minimal?), and during OM (not hot) ?

Depending on how important is backward compatibility (I would say paramount for `CSSStyleRule`?), the idea was to not touch `CSSStyleDeclaration` at all (keep it a map like iterable) because it's backward compatible and needed for other stuff such as style attributes anyway.

But adding a `CSSBlockContent` parent class for all nested group rules (the most important one being `CSSStyleRule`) (note that modifying the `CSSGroupingRule` interface might also works and be less invasive)

```cpp
enum BlockElement {
  CSSStyleDeclaration, // slight preference to return block of declaration, but it could also be each property individually ?
  CSSRule
}

class CSSBlockContent {
  array<BlockElement> elements;
}

class CSSStyleRule : CSSBlockContent {
   CSSStyleAttribute style() // return the current style attribute including all the properties inside the rule but ignoring the interleaving rules
 }
```
For the padding example
```
foo {
  padding-top: 0;
  padding-left: 0;
  & { padding-top: 10px; }
  padding-bottom: 0;
  padding-right: 0;
}
```
Calling `style()` would return `padding: 0px` (which is really what this rule is about), but calling `elements()` would return 
```
[ 
	CSSStyleDeclaration(padding-top: 0, padding-left: 0);
	CSSStyleRule(...);
	CSSStyleDeclaration(padding-bottom: 0px; padding-right: 0)
]
```

**Comment by emilio** [comment]
> during style resolving for order of appearance (minimal?)

Doesn't seem minimal to me fwiw (though happy to be proven wrong). Right now you need to sort _rules_ by source order, and with this change you'd need to sort _declarations_ by source order which can be much more expensive (e.g. with the `all` property or so, but also normally). Not sure that'd fly...

**Comment by emilio** [comment]
(I guess you can still somehow cheat when there's no interleaving but, ugh...)

**Comment by Loirooriol** [comment]
> you'd need to sort declarations by source order

Isn't that already the case due to logical properties? https://drafts.csswg.org/css-logical-1/#box

> this requires implementations to maintain relative order of declarations within a CSS declaration block

**Comment by emilio** [comment]
No, declarations in the same block are stored in the right order you don't need to sort those by source order during styling (unlik rules). That's because within the same declaration block, all of them either match or don't.

**Comment by mdubet** [comment]
> No, declarations in the same block are stored in the right order you don't need to sort those by source order during styling (unlik rules). That's because within the same declaration block, all of them either match or don't.

What is the "right order" to store declarations within a block currently ?

**Comment by emilio** [comment]
The order they were declared. cssom has rules to change the order if needed if you set a logical property that is already present if a physical property could override it.

But the important property is that they either all match or not match, and they all get applied in order, so you don't need to sort amongst them. With your approach you need to potentially split a declaration block to apply things interleaved. It's not impossible to make it fast with some effort I'm sure, but seems non-trivial to me.

**Comment by astearns** [comment]
Are we any closer to agreeing on a solution here? I am not clear whether @mdubet and @emilio are converging on something yet.

**Comment by romainmenke** [comment]
I am looking at the pages on the use counter and there seem to be a lot of false positives. I haven't found a real error here. 🤔 


https://www.jamieoliver.com/

_Selectors for rules before the declaration are for other elements_

```css
.slick-slide {
    float: left;
    height: 100%;
    min-height: 1px;
    [dir=rtl] & {
        float: right
    }

    img {
        display: block
    }

    &.slick-loading img {
        display: none
    }

    display: none; /* here */
    &.dragging img {
        pointer-events: none
    }

    .slick-initialized & {
        display: block
    }

    .slick-loading & {
        visibility: hidden
    }

    .slick-vertical & {
        display: block;
        height: auto;
        border: 1px solid transparent
    }
}
```

https://www.furniturevillage.co.uk/

_Can't nest pseudo element selectors, so would never work anyway_

```css
.resource-loader:before {
    background-image: url("data:image/svg+xml;charset=US-...");
    .ie-browser .resource-loader:before {
        background: none; /* here */
    }

    -o-background-size: 100px;
    background-size: 100px;
    background-position: center center;
    z-index: 7076
}
```


https://www.goat.com/en-nl

_other elements and other properties_

```css
.hideScrollbars {
    &::-webkit-scrollbar {
        -webkit-appearance: none;
        display: none
    }

    -ms-overflow-style: none;
    scrollbar-width: none
}
```

-------

I am unsure if this list was reset when the use counter was refined.

**Comment by LeaVerou** [comment]
I mentioned this in one of the calls, but I figured I should write it here too for posterity: The past couple of months I have experimented with teaching novices CSS nesting from the start, as it's a *much* more straightforward way to explain things than descendant selectors. I did not mention anything about the order of declarations and rules and left it up to them. Almost everyone’s code that I have seen naturally ends up having declarations interleaved with rules. 

My takeaways:
1. So, _so_ glad we ended up with one of the options that allowed this.
2. We _need_ to fix this issue, ASAP. The hidden `@nest` approach outlined above seems to work well, for both authors and implementors. What's the blocker?

@tabatkins nit, but I wonder if `@this` might be a better name? Since it's not really nesting?

**Comment by romainmenke** [comment]
> Almost everyone’s code naturally ends up having declarations interleaved with rules.

Do you have any insights into how many times they interleaved in a way that the order actually mattered?

e.g.

```css
.foo {
  .theme-blue & {
    color: blue; /* never applied if we don't shift up */
  }

  color: green;
}
```

I am expecting this to be extremely rare.

I know that this would only be anecdotal, but I still find it interesting :)

**Comment by kizu** [comment]
@romainmenke I did write my comment without seeing yours, but for

```
.foo {
  .theme-blue & {
    color: blue;
  }

  color: green;
}
```

the color from `.theme-blue` will always override the green one thanks to specificity, so this is a non-issue here, and we can safely do whatever :)

**Comment by romainmenke** [comment]
@kizu thank you, updated to an example with the same specificity :)
Something with conditional at rules might have been a better example.

------

But, yeah I suspect that the use counter isn't tracking the actual problem at hand here.

**Comment by LeaVerou** [comment]
> > Almost everyone’s code naturally ends up having declarations interleaved with rules.
> 
> Do you have any insights into how many times they interleaved in a way that the order actually mattered?

I sadly do not — I only see their code very briefly and synchronously. It's long enough to tell they’re interleaving, but not long enough to carefully examine the dependencies.

**Comment by andruud** [comment]
> We need to fix this issue, ASAP.

+1, if we're going to do this, we should not wait any longer.

> I am looking at the pages on the use counter

(I assume you're talking about the old use-counter?)

Anyway, as previously stated, I added more advanced use-counters a while back, and it’s been in Chrome Stable for a few weeks now.

 - [CSSBareDeclarationShift](https://chromestatus.com/metrics/feature/timeline/popularity/4783) (0.00029%), which triggers when retaining interleaving/trailing declarations in their place (by way of a wrapping `@nest` rule) would make a difference to the result of the cascade. (Yes @kizu, this counter does the right thing in the cases you mentioned).
 - [CSSNestedGroupRuleSpecificity](https://chromestatus.com/metrics/feature/timeline/popularity/4784) (0.00017%), which triggers when `@nest`-wrapping bare declarations within a nested group rule would make a difference to the cascade. Making this move would allow nesting e.g. `@media` in pseudo-element rules, which doesn't work currently since we're wrapping in an `&`-rule.

These numbers are small, and I’m pretty confident that Blink API Owners will approve this change in Chrome, provided that we act quickly, and have clear willingness from Mozilla and Apple to actually carry out the change (@emilio, @fantasai).

The [previous meeting notes](https://github.com/w3c/csswg-drafts/issues/8738#issuecomment-1969468715) suggests that we're waiting for use counter data. I think we can consider this unblocked now.

**Comment by emilio** [comment]
> clear willingness from Mozilla and Apple to actually carry out the change

I'm supportive of this, specially if `@nest` also fixes the subtle issues with nesting media queries and so on on stuff like pseudo selectors.

**Comment by romainmenke** [comment]
Thank you @andruud,

I was checking the old link, I somehow assumed it would update with the new use counter :)
These new numbers are very reassuring!

**Comment by andruud** [comment]
That proposal sounds good.

> without imposing the "basically wrapped in an :is()"

Stating the obvious (?), this means we'll get the regular specificity characteristics as well. E.g. for `div,#foo { span {...} @nest { color:green; } }`, the `@nest` matches `<div>` with the specificity of `div`, not with the specificity of `#foo`.

**Comment by emilio** [comment]
@LeaVerou I'd prefer not to do that, because the nest rule needs to be in the OM, and if you ask for the serialization of only that rule, then we need to serialize the prelude anyways...

**Comment by andruud** [comment]
@emilio Can we not serialize it as bare declarations if it's being serialized using e.g. `cssText` of an outer rule that holds it, but otherwise serialize with the `@nest {}` included?

**Comment by emilio** [comment]
We _can_, but it's a bit confusing / too magic IMO... I don't think we have a precedent for that, and it's weird if you have an item in .cssRules that you otherwise don't see in the serialization or anything... I guess I'm not strongly opposed, but I would prefer to just be consistent with every other CSS rule...

**Comment by andruud** [comment]
OK, I see. Well in that case, a possible additional argument against dropping the `@nest{}` is that it won't always round-trip "structurally", e.g. `div { @nest { color: red; } }` => `div { color: red; }`. They have the same meaning in terms of matching, but CSSOM would reveal different structures.

**Comment by LeaVerou** [comment]
Yeah, I was talking about cases like `.cssText`, where it's serialized as part of an outer rule. I don't have an opinion on what should happen if it's serialized by itself. What I'm trying to avoid is authors getting `@nest` rules that they never specified. Since given an `@nest` rule, it's _far_ more likely that it was implicitly generated, rather than author-specified, I'd rather have serialization follow the principle of least surprise.

**Comment by tabatkins** [comment]
I would prefer to serialize the `@nest` always, because authors can write things like:

```css
.foo {
  color: blue;
  @nest {
    color: red;
  }
}
```

And I'd prefer not to make the choice of whether to serialize or not dependent on whether a *non-`@nest`* rule has been seen yet.

**Comment by emilio** [comment]
Well, but that would have the problem @andruud mentions above, right? If the author _does_ write `@nest`, then the result in the OM is different. And I don't think we want to keep a magic "is implicit" bit around, because that'd be even more magic.

**Comment by emilio** [comment]
Ah, mid-aired with @tabatkins, but we made the same argument effectively :)

**Comment by tabatkins** [comment]
You can't even use a bit, because an author could write

```css
.foo {
  color: blue;
  @media (...) {...}
  color: red;
}
```

And then *delete* the MQ via CSSOM, thus putting the `@nest` back up against the bare declarations. So you'd need to determine it fresh each time.

**Comment by mdubet** [comment]
If we define the @nest as:
- mostly/exclusively used to automatically wrap declarations 
- mostly invisible to user in the serialisation
- not having a specific behaviour (just using the behaviour/specificity of the parent style rule)

Do we actually need to define it ? Wouldn't extending the API of a `CSSGroupingRule` to allow a list of intermixed `CSSStyleDeclaration` and `CSSRule` works ?

**Comment by emilio** [comment]
Kind of? How would you expose that? What would `CSSGroupingRule.cssRules` do? What would happen if you remove the middle rule in @tabatkins's example above? Would the CSSStyleDeclarations get magically merged? Seems less straight-forward / more questions, at the very least...

**Comment by LeaVerou** [comment]
> I would prefer to serialize the `@nest` always, because authors can write things like:
> 
> ```css
> .foo {
>   color: blue;
>   @nest {
>     color: red;
>   }
> }
> ```
> 
> And I'd prefer not to make the choice of whether to serialize or not dependent on whether a _non-`@nest`_ rule has been seen yet.

Sure they _could_, but we don’t design APIs based on the 1% of cases. Far better to have reasonable behavior in the 99% of cases and magic in the 1% than the other way around, even if the former seems more magic.

Also @mdubet makes a good point: is there value in authors actually writing `@nest`? What if they can't, and it basically becomes a CSSOM implementation detail?

**Comment by mdubet** [comment]
@emilio Either `CSSGroupingRule.cssRules` returns a list of `CSSRule` + `CSSStyleDeclaration` or we use a new `CSSGroupingRule.elements` or something else.

It's really the same idea, except that we use `CSSStyleDeclaration` instead of a potential `@nest`. But it has the advantage of not being explicitly writeable by a user (you can't write distinct consecutive CSSStyleDeclaration blocks in CSS - but you can if you really want with CSSOM).

(For the example above, merging the two `CSSStyleDeclaration` would makes more sense indeed)

**Comment by tabatkins** [comment]
> Sure they could, but we don’t design APIs based on the 1% of cases. Far better to have reasonable behavior in the 99% of cases and magic in the 1% than the other way around, even if the magic case in the former seems more magic.

Basically my opinion is that the more magic we have for this, the worse it is. There's a minimum amount of magic we need for usability's sake (automatically wrapping the declaration in this new rule when parsing), but beyond that I strongly prefer for everything to be as normal and unsurprising as possible. Every bit of magic is another sharp corner we have to be aware of and design around, forever, vs something predictable and boring.

Importantly, pay attention to the use-case you're talking about - when does an author serialize a CSS rule and manually examine the results? That seems... super rare to me. It's the sort of thing I, personally, have only ever done to debug things. Putting aside the question of whether it's better to serialize the implicit rule or not, does this case occur often enough that it's worth doing *anything* custom at all? In my opinion, definitely not.

> [talk about using CSSStyleDeclaration]

Same point, really - is it worth doing *anything* more complicated than just "here's a new type of at-rule, it acts like you'd expect"?

**Comment by LeaVerou** [comment]
I don't understand why you think it's acceptable to magically _add_ `@nest`, but not magically _remove_ it. They seem like two sides of the same coin to me, and if anything doing one without the other is inconsistent. Having `@nest` magically appear in my code when I try to roundtrip it seems way weirder than seeing a manually authored `@nest` disappear in the same case. And it's still unclear to me what is the value of allowing authors to write `@nest` at all. If there is none, then the case of an author writing `@nest` and seeing it disappear if they try to roundtrip will be _exceedingly_ rare. 

> Importantly, pay attention to the use-case you're talking about - when does an author serialize a CSS rule and manually examine the results? 

I believe certain types of dev tools do this to at least some degree.

> There's a minimum amount of magic we need for usability's sake (automatically wrapping the declaration in this new rule when parsing), but beyond that I strongly prefer for everything to be as normal and unsurprising as possible. 

Are you saying that the expected frequency with which something is encountered doesn't factor in the API design choices you make? 

> Every bit of magic is another sharp corner we have to be aware of and design around, forever, vs something predictable and boring.

Wait, I thought we were talking about author ergonomics, when did it become about our convenience?

**Comment by astearns** [comment]
In the interests of resolving this 90-comment issue, would it be OK to open a new issue on the details of `@nest`, or do we need to figure it all out to make any progress?

**Comment by LeaVerou** [comment]
> In the interests of resolving this 90-comment issue, would it be OK to open a new issue on the details of `@nest`, or do we need to figure it all out to make any progress?

I think we can probably resolve on the general proposal, then discuss serialization in a separate issue.

**Comment by mdubet** [comment]
> Same point, really - is it worth doing _anything_ more complicated than just "here's a new type of at-rule, it acts like you'd expect"?

More complicated for browser engine indeed, but less complicated for CSS users (no need to learn about a @nest rule which will maybe create some confusion with the nesting selector & ..etc)

**Comment by emilio** [comment]
> More complicated for browser engine indeed, but less complicated for CSS users

Not sure I agree. This requires to expand the API surface of CSSOM quite a bit, or changing APIs in ways that are confusing / inconsistent (like not returning rules from .cssRules), or unclear...

If you're not inspecting stylesheets via script, it doesn't matter either way. If you do, then IMO the `@nest` rule is easier to deal with than something new / more complicated. It just works like a style rule.

Given that, and that this is somewhat of a time bomb which we probably want to fix sooner rather than later, I tend to think that the simplest / less magical solution is best.

 * It's implementable in a very straight-forward / performant way in all engines.
 * It raises less open / follow-up questions.
 * It's (IMO) not worse for authors than the current state of things or alternatives.

@tabatkins regarding the proposal, I'm assuming:

 * `@nest` doesn't need to be a grouping rule (though I guess it doesn't matter much either way).
 * With "style rules" in your proposal you mean style declarations, presumably, right?

**Comment by andruud** [comment]
Fully agree with @emilio.

I _initially_ thought that we should serialize back to bare declarations for outer `cssText` since it's shorter, nicer, and basically "why not". But the round-trip issues and other edge cases convinced me the other way.

@LeaVerou You said that we should consider another name for `@nest`, would that mitigate the situation? If the name better explains what it does/is, maybe its presence will feel less offensive? `@this`? `@self`? `@wart`?

> changing APIs in ways that are confusing / inconsistent (like not returning rules from .cssRules)

Yes, let's avoid this, please. Getting creative about the OM structure is even more intrusive than getting creative about its serialization.

**Comment by tabatkins** [comment]
To reiterate my points a bit from the call:

* Fixing this problem *in any reasonable way* will affect (positively) approximately all CSS authors (by making properties in MQs/etc work better), and unblock CSS from evolving in certain ways that we want it to (adding mixins). 
* The issues brought up in the latter half of this thread are exclusively about the CSSOM, specifically reading and modifying it. I, personally, have never *once* read or modified the CSSOM in my entire webdev career; I've only ever done so to write WPTs.
* So, to a first approximation, these issues affect nobody. ^_^ More precisely, the only audience I know of that actually reads/modifies the CSSOM in non-trivial ways (more than just, say, serializing an entire stylesheet and putting it somewhere else) are *CSS tool authors*. I'd estimate there to be *a few hundred* people in this category, across the entire world; a thousand on the extreme end of the possible range. This audience is also going to be significantly more experienced with coding in general.
* None of the issues listed here are about a use-case being *difficult* or *impossible*, which is usually what we can use to justify solving a use-case for such a small audience. They are convenience features.
* And the convenience itself is arguable in most cases. They're mostly about making the OM more *consistent* in some ways, but every single one comes at the cost of making it *less consistent* in other ways.

So in summary, past the necessary complexity/magic that's at the core of the issue (grouping interleaved declarations in some way so they don't merge with the initial set of decls), *any* further complexity we add is, as far as I can tell, solely to add minor convenience to some use-cases for a miniscule audience of already-experienced developers. Thus, none of them justify the spec and impl complexity cost they would impose.

I'm more than happy to be shown wrong on this regard - if you have an example of some change to [my base proposal](https://github.com/w3c/csswg-drafts/issues/8738#issuecomment-2057868087) that would either give the "CSS tool authors" audience a *large* benefit (especially if it unblocks something that is currently hard or impossible), or that would give the "CSS general audience" even a mild benefit, or that would give some third audience I'm not aware of a benefit that is appropriately weighted against its size, then I'd joyfully consider it.

**Comment by astearns** [comment]
But let’s not make this a permathread to go over all the alternatives and options. This issue was specifically about changing the “shifting up” behavior, which we have resolved to undo (to great author benefit). Please open new issues on specific problems or alternative solutions so that we can have more focused discussion about them.

@mdubet could you open an issue on the idea of using multiple CSSStyleDeclarations instead of a new `@nest` rule?

@LeaVerou would it make sense to open a new issue to explore how far we could go in making things browser-internal and not exposed to developers?

@kbabbitt if you think it is possible to solve the setProperty() problem while keeping the non-shifting behavior, please open a new issue

(and anything else I have missed, please open a new issue)

**Comment by yisibl** [comment]
@LeaVerou It looks like `@nest` [has been removed](https://github.com/w3c/csswg-drafts/commit/7d8f87c8b8de9b7dd64719aeac18004ea76fdad0) from the specification again right? This history I think needs an article detailing it.
